File: src/main/java/com/github/ccguyka/showupdates/filter/FilterExcludedArtifacts.java
package com.github.ccguyka.showupdates.filter;

import static java.util.stream.Collectors.toList;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.function.Predicate;
import org.apache.maven.artifact.Artifact;
import org.apache.maven.artifact.versioning.ArtifactVersion;

import com.google.common.collect.Lists;

/**
 * @todo Write tests for FilterExcludedArtifacts class.
 */
public class FilterExcludedArtifacts {

    private final List<String> excludes;

    public FilterExcludedArtifacts(final String... excludes) {
        this.excludes = Lists.newArrayList(excludes);
    }

    public Map<Artifact, List<ArtifactVersion>> filter(final Map<Artifact, List<ArtifactVersion>> updates) {
        final Map<Artifact, List<ArtifactVersion>> filteredExcludedArtifacts = new HashMap<>();
        for (final Entry<Artifact, List<ArtifactVersion>> update : updates.entrySet()) {
            filteredExcludedArtifacts.put(update.getKey(), filterExcludedArtifacts(update.getValue()));
        }

        return filteredExcludedArtifacts;
    }

    private List<ArtifactVersion> filterExcludedArtifacts(final List<ArtifactVersion> update) {
        return update.stream().filter(filterExcludedArtifact()).collect(toList());
    }

    private Predicate<? super ArtifactVersion> filterExcludedArtifact() {
        return artifact -> Lists.newArrayList(excludes).stream()
                .noneMatch(exclude -> artifact.toString().contains(exclude));
    }
}

File: src/main/kotlin/com/github/ccguyka/showupdates/filter/FilterExcludedArtifacts.kt
package com.github.ccguyka.showupdates.filter

import com.google.common.collect.Lists
import org.apache.maven.artifact.Artifact
import org.apache.maven.artifact.versioning.ArtifactVersion
import java.util.*
import java.util.function.Predicate
import java.util.stream.Collectors

/**
 * Filter out artifacts of which the version contains one of the excludes.
 */
class FilterExcludedArtifacts(vararg excludes: String) {

    private val excludes: List<String>

    fun filter(updates: Map<Artifact, List<ArtifactVersion>>): Map<Artifact, List<ArtifactVersion>> {
        val filteredExcludedArtifacts: MutableMap<Artifact, List<ArtifactVersion>> = HashMap()
        for ((key, value) in updates) {
            filteredExcludedArtifacts[key] = filterExcludedArtifacts(value)
        }
        return filteredExcludedArtifacts
    }

    private fun filterExcludedArtifacts(update: List<ArtifactVersion>): List<ArtifactVersion> {
        return update.stream().filter(filterExcludedArtifact()).collect(Collectors.toList())
    }

    private fun filterExcludedArtifact(): Predicate<in ArtifactVersion> {
        return Predicate { artifact: ArtifactVersion ->
            Lists.newArrayList(excludes).stream()
                    .noneMatch { s: String? -> artifact.toString().contains(s!!) }
        }
    }

    init {
        this.excludes = Lists.newArrayList(*excludes)
    }
}
